File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/page/modules/custommodules/GameModule.kt
package com.crescentflare.slicinggame.page.modules.custommodules

import android.content.Context
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.components.containers.GameContainerView
import com.crescentflare.slicinggame.components.utility.ImageSource
import com.crescentflare.slicinggame.infrastructure.events.AppEvent
import com.crescentflare.slicinggame.page.activities.PageActivity
import com.crescentflare.slicinggame.page.modules.PageModule

/**
 * Custom module: handles custom logic on the game page
 */
class GameModule: PageModule {

    // --
    // Static: inflatable integration
    // --

    companion object {

        val inflatable: JsonInflatable = object : JsonInflatable {

            override fun create(context: Context): Any {
                return GameModule()
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is GameModule) {
                    // Apply component reference
                    val layoutBinder = (parent as? PageActivity)?.binder
                    val component = layoutBinder?.findByReference(mapUtil.optionalString(attributes, "gameContainer", null) ?: "")
                    obj.gameContainer = component as? GameContainerView

                    // Apply random backgrounds
                    val randomBackgrounds = mutableListOf<ImageSource>()
                    mapUtil.optionalObjectList(attributes, "randomBackgrounds").forEach {
                        val randomBackground = ImageSource.fromValue(it)
                        if (randomBackground != null) {
                            randomBackgrounds.add(randomBackground)
                        }
                    }
                    obj.randomBackgrounds = randomBackgrounds
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == GameModule::class
            }

        }

    }


    // --
    // Members
    // --

    override val handleEventTypes = listOf("game")
    private var currentBackground = 0
    private var shuffledBackgrounds: List<ImageSource> = emptyList()


    // --
    // Configurable values
    // --

    var gameContainer: GameContainerView? = null
        set(gameContainer) {
            if (gameContainer !== field && currentBackground < shuffledBackgrounds.size) {
                gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
            }
            field = gameContainer
        }

    var randomBackgrounds: List<ImageSource> = emptyList()
        set(randomBackgrounds) {
            if (randomBackgrounds != field && randomBackgrounds.isNotEmpty()) {
                currentBackground = 0
                shuffledBackgrounds = randomBackgrounds.shuffled()
                gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
            }
            field = randomBackgrounds
        }


    // --
    // Lifecycle
    // --

    override fun onCreate(activity: PageActivity) {
        // No implementation
    }

    override fun onResume() {
        // No implementation
    }

    override fun onPause() {
        // No implementation
    }

    override fun onDestroy() {
        // No implementation
    }

    override fun onLowMemory() {
        // No implementation
    }

    override fun onBackPressed(): Boolean {
        return false // No custom handling for the back button
    }


    // --
    // Event handling
    // --

    override fun handleEvent(event: AppEvent, sender: Any?): Boolean {
        if (event.type == "game") {
            when (event.name) {
                "reset" -> {
                    gameContainer?.resetSlices()
                    if (shuffledBackgrounds.isNotEmpty()) {
                        currentBackground = (currentBackground + 1) % shuffledBackgrounds.size
                        gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
                    }
                }
            }
            return true
        }
        return false
    }

}

File: slicing-ios/SlicingGame/Page/Modules/CustomModules/GameModule.swift
//
//  AlertModule.swift
//  Custom module: handles custom logic on the game page
//

import UIKit
import JsonInflator

class GameModule: PageModule {
    
    // --
    // MARK: Members
    // --
    
    let handleEventTypes = ["game"]
    private var currentBackground = 0
    private var shuffledBackgrounds = [ImageSource]()

    
    // --
    // MARK: Inflator integration
    // --
    
    class func inflatable() -> JsonInflatable {
        return InflatorClass()
    }
    
    private class InflatorClass: JsonInflatable {
        
        func create() -> Any {
            return GameModule()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let gameModule = object as? GameModule {
                // Apply component reference
                let layoutBinder = (parent as? PageViewController)?.binder
                let component = layoutBinder?.findByReference(convUtil.asString(value: attributes["gameContainer"]) ?? "")
                gameModule.gameContainer = component as? GameContainerView
                
                // Apply random backgrounds
                var randomBackgrounds = [ImageSource]()
                for randomBackground in attributes["randomBackgrounds"] as? [Any] ?? [] {
                    if let image = ImageSource.fromValue(value: randomBackground) {
                        randomBackgrounds.append(image)
                    }
                }
                gameModule.randomBackgrounds = randomBackgrounds
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == GameModule.self
        }
        
    }


    // --
    // MARK: Initialization
    // --
    
    init() {
        // No implementation
    }


    // --
    // MARK: Configurable values
    // --
    
    var gameContainer: GameContainerView? {
        didSet {
            if gameContainer != oldValue && currentBackground < shuffledBackgrounds.count {
                gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
            }
        }
    }
    
    var randomBackgrounds = [ImageSource]() {
        didSet {
            if randomBackgrounds != oldValue && randomBackgrounds.count > 0 {
                currentBackground = 0
                shuffledBackgrounds = randomBackgrounds.shuffled()
                gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
            }
        }
    }


    // --
    // MARK: Lifecycle
    // --

    func didCreate(viewController: PageViewController) {
        // No implementation
    }
    
    func didResume() {
        // No implementation
    }
    
    func didPause() {
        // No implementation
    }
    
    func willDestroy() {
        // No implementation
    }
    

    // --
    // MARK: Event handling
    // --

    func handleEvent(_ event: AppEvent, sender: Any?) -> Bool {
        if event.type == "game" {
            switch event.name {
            case "reset":
                gameContainer?.resetSlices()
                if shuffledBackgrounds.count > 0 {
                    currentBackground = (currentBackground + 1) % shuffledBackgrounds.count
                    gameContainer?.backgroundImage = shuffledBackgrounds[currentBackground]
                }
            default:
                break
            }
            return true
        }
        return false
    }

}
